pkg/gauges/contracts/gauges/PremintedGauge.sol

```diff
@@ -0,0 +1,137 @@
+        // be used prior to the start of the first emission period
+        uint256 rate = tokenAdmin.rate();
+        require(rate != 0);
+
+        _rate = rate;
+        _period = _currentPeriod();
+        _startEpochTime = tokenAdmin.startEpochTimeWrite();
+    }
+
+    function checkpoint() external nonReentrant returns (bool) {
+        // TODO: Add a guard to checkpointing.
+        uint256 lastPeriod = _period;
+        uint256 currentPeriod = _currentPeriod();
+
+        if (lastPeriod < currentPeriod) {
+            _gaugeController.checkpoint_gauge(address(this));
+
+            uint256 rate = _rate;
+            uint256 newEmissions = 0;
+            lastPeriod += 1;
+            uint256 nextEpochTime = _startEpochTime + _RATE_REDUCTION_TIME;
+            for (uint256 i = lastPeriod; i < lastPeriod + 255; ++i) {
+                if (i > currentPeriod) break;
+
+                uint256 periodTime = i * 1 weeks;
+                uint256 periodEmission = 0;
+                uint256 gaugeWeight = _gaugeController.gauge_relative_weight(address(this), periodTime);
+
+                if (nextEpochTime >= periodTime && nextEpochTime < periodTime + 1 weeks) {
+                    // If the period crosses an epoch, we calculate a reduction in the rate
+                    // using the same formula as used in `ERC20CRV`. We perform the calculation
+                    // locally instead of calling to `ERC20CRV.rate()` because we are generating
+                    // the emissions for the upcoming week, so there is a possibility the new
+                    // rate has not yet been applied.
+
+                    // Calculate emission up until the epoch change
+                    periodEmission = (gaugeWeight * rate * (nextEpochTime - periodTime)) / 10**18;
+                    // Action the decrease in rate
+                    rate = (rate * _RATE_DENOMINATOR) / _RATE_REDUCTION_COEFFICIENT;
+                    // Calculate emission from epoch change to end of period
+                    periodEmission += (gaugeWeight * rate * (periodTime + 1 weeks - nextEpochTime)) / 10**18;
+
+                    _rate = rate;
+                    _startEpochTime = nextEpochTime;
+                    nextEpochTime += _RATE_REDUCTION_TIME;
+                } else {
+                    periodEmission = (gaugeWeight * rate * 1 weeks) / 10**18;
+                }
+
+                //log PeriodEmission(period_time, period_emission)
+                newEmissions += periodEmission;
+            }
+
+            _period = currentPeriod;
+            _emissions += newEmissions;
+
+            if (newEmissions > 0 && !isKilled) {
+                _minter.mint(address(this));
+                _postMintAction(newEmissions);
+            }
+        }
+
+        return true;
+    }
+
+    function _currentPeriod() internal view returns (uint256) {
+        return (block.timestamp / 1 weeks) - 1;
+    }
+
```

Comment:
We seem to be missing the `future_epoch_time()` function

Reply:
This doesn't seem to follow. Why would we want this function?